dusk-labyrinth.lic

```diff
@@ -7,74 +7,208 @@ custom_require.call %w[common common-items events]
 class DuskLab
   include DRC
   include DRCI
+
   def initialize
+
+    arg_definitions = [
+      [
+        { name: 'debug', regex: /debug/i, optional: true, description: 'Enable debug output' },
+      ]
+    ]
+
+    args = parse_args(arg_definitions)
+    $debug_mode_dl = UserVars.dusk_labyrinth_debug || args.debug || false
+
     @settings = get_settings
     @labyrinth_data = @settings.duskruin['labyrinth']
+
     loot_container = @labyrinth_data['loot_container']
     pet_noun = @labyrinth_data['pet_noun']
     redeem_scrip = @labyrinth_data['redeem_scrip']
+
+    # Map of directions to help know where we've come from and how to go back.
+    @reverse_direction_map = {
+      'e' => 'w',
+      'w' => 'e',
+      's' => 'n',
+      'n' => 's',
+      'ne' => 'sw',
+      'sw' => 'ne',
+      'nw' => 'se',
+      'se' => 'nw',
+      'east' => 'west',
+      'west' => 'east',
+      'south' => 'north',
+      'north' => 'south',
+      'northeast' => 'southwest',
+      'southwest' => 'northeast',
+      'northwest' => 'southeast',
+      'southeast' => 'northwest'
+    }
+
+    # Be notified when certain events happen, such as seeing a pet move or the maze reshuffle.
     Flags.add('five-minute-warning', 'You only have about 5 minutes left in the labyrinth!')
     Flags.add('done', 'A surly Dwarf stomps in and glowers at you in the dim gloom')
     Flags.add('no-pet', "A small #{pet_noun} scurries #{Regexp.new('(?<pet_dir>\w+)')}!")
     Flags.add('pet', "You notice a small #{pet_noun} scurrying around the area")
     Flags.add('caught', 'You pick them both up, claiming your new pet')
-    @prev_dir = nil
-    @dir_to_prev_dir = {
-      'northeast' => 'southwest',
-      'southwest' => 'northeast',
-      'northwest' => 'southeast',
-      'southeast' => 'northwest',
-      'north' => 'south',
-      'south' => 'north',
-      'east' => 'west',
-      'west' => 'east',
-      'up' => 'down', # up and down dont appear in the maze, but are used for testing
-      'down' => 'up'
-    }
+    Flags.add('maze-shuffled', 'no longer certain of your directions')
+
+    # Let's do this!
+    init_maze
     stow_hands
     main(loot_container) until done?
     redeem if redeem_scrip
   end
 
-  def redeem
-    while bput('get my bloodscrip', 'You get', 'What were') == 'You get'
-      bput('redeem my bloodscrip', 'You quickly pocket')
-    end
+  # Used at beginning of script and any time the maze shuffles.
+  def init_maze
+    echo("*** initializing map data") if $debug_mode_dl
+    # Maze data for the current room.
+    @current_room = nil
+    # Maze data for the previous room.
+    @previous_room = nil
+    # Direction we last moved to go from "previous_room" to "current_room".
+    @forward_move = nil
+    @reverse_move = nil
```

Comment:
I realize this is a refactor and may not be your original script, so feel free to ignore comments about things that you just moved, I'm just giving some off the cuff thoughts. 

There's no need to initialize variables like this, they effectively exist as soon as the code references them anywhere, and are already set to nil. 

```
irb(main):001:0> class Test
irb(main):002:1>   def initialize
irb(main):003:2>     puts @not_mentioned.nil?
irb(main):004:2>   end
irb(main):005:1> end
=> :initialize
irb(main):006:0> Test.new
true
```

If you want a place to call out the variables you'll be using you could use [attribute accessors](https://www.rubyguides.com/2018/11/attr_accessor/) which also gives you some debug access to state if you hang onto the created object from this class.

Reply:
Thanks for the insightful and helpful coaching :) I am learning Ruby _as I refactor_ Lich scripts, hehe

Good to know that variables don't have to be pre-defined, and thanks for the tip about attribute accessors.

In this very specific case, those variable assignments are part of the `init_maze` function defined at line 65 which is called at script start but also anytime the maze reshuffles. It's when the maze reshuffles that they needed to be set back to nil.

Reply:
Ah yeah, my brain saw init_maze and turned it into initialize, I thought this was inside the constructor, ignore the comment in general, but yes the call on 58 doesn't do anything and a more accurate name might be `reset_maze_data` or something similar.